### PR TITLE
PR for dkr/on-demand-split/per-tenant-remote-sync

### PR DESCRIPTION
**Pull request by problame**
With these changes, `test_remote_storage.py` passes on my machine.

I advise to review the commits one by one.

### CODE REVIEW

**Review by LizardWizzard — COMMENTED** [review]
Looks good to me, left minor comments

File: pageserver/src/tenant_mgr.rs [hunk 1]
```diff
@@ -114,19 +124,50 @@ fn load_local_tenant(
     conf: &'static PageServerConf,
     tenant_path: &Path,
     remote_storage: Option<GenericRemoteStorage>,
-) -> anyhow::Result<()> {
+) -> anyhow::Result<Option<()>> {
+    if !tenant_path.is_dir() {
+        anyhow::bail!("tenant_path is not a directory: {tenant_path:?}")
+    }
+
+    if tenant_path.is_empty() {
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant_mgr.rs:132]: Woops, good catch.

File: pageserver/src/tenant.rs [hunk 2]
```diff
@@ -517,29 +540,43 @@ impl Tenant {
...
         }
 
+        std::fs::remove_file(&marker_file)
+            .with_context(|| format!("unlink attach marker file {}", marker_file.display()))?;
+        crashsafe::fsync(
+            marker_file
+                .parent()
+                .expect("marker file is created in tenant directory"),
+        )
+        .with_context(|| format!("fsync tenant directory after unlinking attach marker file"))?;
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant.rs:578]: Nope, caller should be responsible for attaching tenant_id.

File: pageserver/src/tenant_mgr.rs [hunk 3]
```diff
@@ -114,19 +124,50 @@ fn load_local_tenant(
...
-    // state.
-    let tenant = Tenant::spawn_load(conf, tenant_id, remote_storage)?;
+    let tenant = if conf.tenant_attaching_mark_file_path(&tenant_id).exists() {
+        info!("tenant {tenant_id} has attaching mark file, resuming its attach operation");
+        if let Some(remote_storage) = remote_storage {
+            Tenant::spawn_attach(conf, tenant_id, &remote_storage)?
+        } else {
+            warn!("tenant {tenant_id} has attaching mark file, but pageserver has no remote storage configured");
+            // XXX there should be a constructor to make a broken tenant
+            // TODO should we use Tenant::load_tenant_config() here?
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant_mgr.rs:151]: Can there even be a config on a tenant that is attaching?

File: pageserver/src/http/routes.rs [hunk 4]
```diff
@@ -404,10 +413,12 @@ async fn tenant_status(request: Request<Body>) -> Result<Response<Body>, ApiErro
             current_physical_size += timeline.get_physical_size();
         }
 
+        let state = tenant.current_state();
         let tenant_info = TenantInfo {
             id: tenant_id,
-            state: tenant.current_state(),
+            state,
             current_physical_size: Some(current_physical_size),
+            has_in_progress_downloads: Some(state == TenantState::Attaching),
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/http/routes.rs:421]: Did you put this comment in the wrong place?
To me, it makes sense to have it in `build_timeline_info_common` where we construct `TimelineInfo`, much less so here where we construct `TenantInfo`.

**Comment by problame** [comment]
Addressed the review comments, some conversations are still unresolved, though.

File: pageserver/src/http/routes.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/http/routes.rs:424]: Yeah, you're right, sorry, confused has_inprogress_downloads with awaits_download in timeline

This is the relevant place to discuss it I think https://github.com/neondatabase/neon/pull/2820/files#r1023151939

File: pageserver/src/http/routes.rs [hunk 5]
```diff
@@ -143,6 +148,10 @@ fn build_timeline_info_common(timeline: &Arc<Timeline>) -> anyhow::Result<Timeli
 
         state,
 
+        // XXX bring back tracking of downloads per timeline, or, introduce
+        // an 'Attaching' state for the timeline and get rid of this field.
```
[the remark this answers is not in the document]
Comment (LizardWizzard) [reply pageserver/src/http/routes.rs:152]: I would say that this is out of scope for this PR. With proper ondemand landed we might not need this at all

File: libs/pageserver_api/src/models.rs [hunk 6]
```diff
@@ -150,6 +150,7 @@ pub struct TenantInfo {
     pub id: TenantId,
     pub state: TenantState,
     pub current_physical_size: Option<u64>, // physical size is only included in `tenant_status` endpoint
+    pub has_in_progress_downloads: Option<bool>,
```
[the remark this answers is not in the document]
Comment (problame) [reply libs/pageserver_api/src/models.rs]: They are already there, I think.
Note this change is not targeting `main` but @LizardWizzard 's https://github.com/neondatabase/neon/pull/2785

File: pageserver/src/http/routes.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/http/routes.rs:152]: Yeah, I think we'll get rid of this field with proper ondemand PR

File: pageserver/src/tenant_mgr.rs [hunk 7]
```diff
@@ -79,16 +80,25 @@ pub fn init_tenant_mgr(
...
-                    number_of_tenants += 1;
-                    if let Err(e) =
-                        load_local_tenant(conf, &tenant_dir_path, remote_storage.clone())
-                    {
-                        error!(
+                    match load_local_tenant(conf, &tenant_dir_path, remote_storage.clone()) {
+                        Ok(Some(_)) => number_of_tenants += 1,
+                        Ok(None) => {
+                            // This case happens if we crash during attach before creating the attach marker file
+                            if let Err(e) = std::fs::remove_dir(&tenant_dir_path) {
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant_mgr.rs:87]: The marker file is within the tenant_dir, and the logic added in this PR uses it to resume the `attach` operation on the tenant, regardless of how far the attach operation progressed. Can't see a problem with that. Can you rephrase?

[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant_mgr.rs:87]: > I think, there should be one place in the startup logic, that does this.

Your code snippet is from `main`.
That code is all gone in https://github.com/neondatabase/neon/pull/278.
This PR targets #2785 , not `main`.

File: pageserver/src/http/routes.rs [hunk 8]
```diff
@@ -264,15 +274,16 @@ async fn timeline_detail_handler(request: Request<Body>) -> Result<Response<Body
     check_permission(&request, Some(tenant_id))?;
 
     let timeline_info = async {
-        let timeline = tokio::task::spawn_blocking(move || {
-            tenant_mgr::get_tenant(tenant_id, true)?.get_timeline(timeline_id, false)
-        })
-        .await
-        .map_err(|e: JoinError| ApiError::InternalServerError(e.into()))?;
+        let tenant = tenant_mgr::get_tenant(tenant_id, true).map_err(ApiError::NotFound)?;
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/http/routes.rs:277]: Good catch, this wasn't intentional!

**Comment by problame** [comment]
Rebased onto @LizardWizzard 's latest rebase, and addressed the review comments.

**Review by hlinnaka — APPROVED** [review]

**Review by LizardWizzard — COMMENTED** [review]
Looks good to me, minor nitpicks
